ruby/ql/lib/codeql/ruby/ast/Pattern.qll

```diff
@@ -97,3 +97,302 @@ class TuplePattern extends Pattern, TTuplePattern {
+    or
+    pred = "getSuffix" and result = this.getSuffix()
+  }
+}
+
+/**
+ * A hash pattern, for example:
+ * ```rb
+ * in {}
+ * in { a: 1 }
+ * in { a: 1, **rest }
+ * in { a: 1, **nil }
+ * in Node{ label: , children: [] }
+ * ```
+ */
+class HashPattern extends CasePattern, THashPattern {
+  private Ruby::HashPattern g;
+
+  HashPattern() { this = THashPattern(g) }
+
+  /** Gets the class this pattern matches objects against, if any. */
+  ConstantReadAccess getClass() { toGenerated(result) = g.getClass() }
+
+  private Ruby::KeywordPattern keyValuePair(int n) { result = g.getChild(n) }
+
+  /** Get the key of the `n`th pair. */
+  StringlikeLiteral getKey(int n) { toGenerated(result) = keyValuePair(n).getKey() }
+
+  /** Get the value of the `n`th pair. */
+  CasePattern getValue(int n) { toGenerated(result) = keyValuePair(n).getValue() }
+
+  /** Get the value for a given key name. */
+  CasePattern getValueByKey(string key) {
+    exists(int i | key = this.getKey(i).getValueText() and result = this.getValue(i))
+  }
+
+  /** Get the variable of the keyword rest token, if any, i.e. `name` in `**name`. */
+  LocalVariableWriteAccess getRestVariable() {
+    toGenerated(result) =
+      max(int i | | g.getChild(i) order by i).(Ruby::HashSplatParameter).getName()
+  }
+
+  /**
+   * Holds if this pattern is terminated by `**nil` indicating that the pattern does not permit
+   * any unmatched remaining pairs.
+   */
+  predicate allowUnmatchedElements() { g.getChild(_) instanceof Ruby::HashSplatNil }
+
+  final override string getAPrimaryQlClass() { result = "HashPattern" }
+
+  final override string toString() { result = "{ ..., ** }" }
+
+  final override AstNode getAChild(string pred) {
+    result = super.getAChild(pred)
+    or
+    pred = "getClass" and result = this.getClass()
+    or
+    pred = "getKey" and result = this.getKey(_)
+    or
+    pred = "getValue" and result = this.getValue(_)
+    or
+    pred = "getRestVariable" and result = this.getRestVariable()
+  }
+}
+
+/**
+ * A composite pattern matching one of the given sub-patterns, for example:
+ * ```rb
+ * in 1 | 2 | 3
+ * ```
+ */
+class AlternativePattern extends CasePattern, TAlternativePattern {
+  private Ruby::AlternativePattern g;
+
+  AlternativePattern() { this = TAlternativePattern(g) }
+
+  /** Gets the `n`th alternative of this pattern. */
+  CasePattern getAlternative(int n) { toGenerated(result) = g.getAlternatives(n) }
+
+  /** Gets an alternative of this pattern. */
+  CasePattern getAnAlternative() { result = this.getAlternative(_) }
+
+  final override string getAPrimaryQlClass() { result = "AlternativePattern" }
+
+  final override string toString() { result = "... | ..." }
+
+  final override AstNode getAChild(string pred) {
+    result = super.getAChild(pred)
+    or
+    pred = "getAlternative" and result = this.getAlternative(_)
+  }
+}
+
+/**
+ * A pattern matching with a rename into specified local variable, for example `Integer => a`
```

Comment:
This sentence is a little hard to understand. Is my suggestion correct, and do you think it's any clearer?
```suggestion
 * A pattern match that binds to the specified local variable, for example `Integer => a`
```